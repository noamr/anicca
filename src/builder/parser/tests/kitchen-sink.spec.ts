import {parse} from 'yaml'
const {parseKal} = require('../index')

describe('kitchen sink', () => {
    it('views with clones', () => {
        expect(parseKal(parse(`
view someView:
    '#some-id':
        content: someValue & hello
        on mousedown:
            - dispatch someEvent to someController
            - prevent default
            - stop propagation
        on click:
            dispatch hello to someOtherController

    '#some-clone':
        for [id, value] in myMap:
            .something:
                content: value
                attribute class: \`\${id}-class\`

`))).toMatchSnapshot()
    })
    it('timers', () => {
        expect(parseKal(parse(`
controller someController:
    root:
        a:
            on someEvent:
                goto b
        b:
            after 30 seconds:
                goto a
        c:
            after a+5 minutes when condition==0:
                goto a
`))).toMatchSnapshot()
    })
})
