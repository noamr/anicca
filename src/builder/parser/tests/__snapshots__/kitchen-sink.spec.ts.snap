// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kitchen sink timers 1`] = `
Array [
  Object {
    "name": "someController",
    "rootState": Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "actions": Array [
                Object {
                  "target": "b",
                  "type": "Goto",
                },
              ],
              "event": "someEvent",
              "type": "Transition",
            },
          ],
          "default": undefined,
          "name": "a",
          "onEntry": Array [],
          "onExit": Array [],
          "type": "State",
        },
        Object {
          "children": Array [
            Object {
              "actions": Array [
                Object {
                  "target": "a",
                  "type": "Goto",
                },
              ],
              "timeout": Object {
                "args": Array [
                  Object {
                    "$primitive": 30,
                  },
                  Object {
                    "$primitive": 1,
                  },
                ],
                "op": "mult",
              },
              "type": "Transition",
            },
          ],
          "default": undefined,
          "name": "b",
          "onEntry": Array [],
          "onExit": Array [],
          "type": "State",
        },
        Object {
          "children": Array [
            Object {
              "actions": Array [
                Object {
                  "target": "a",
                  "type": "Goto",
                },
              ],
              "condition": Object {
                "args": Array [
                  Object {
                    "$ref": "condition",
                  },
                  Object {
                    "$primitive": 0,
                  },
                ],
                "op": "eq",
              },
              "timeout": Object {
                "args": Array [
                  Object {
                    "args": Array [
                      Object {
                        "$ref": "a",
                      },
                      Object {
                        "$primitive": 5,
                      },
                    ],
                    "op": "add",
                  },
                  Object {
                    "$primitive": 60000,
                  },
                ],
                "op": "mult",
              },
              "type": "Transition",
            },
          ],
          "default": undefined,
          "name": "c",
          "onEntry": Array [],
          "onExit": Array [],
          "type": "State",
        },
      ],
      "default": "a",
      "name": "root",
      "onEntry": Array [],
      "onExit": Array [],
      "type": "State",
    },
    "type": "Controller",
  },
]
`;

exports[`kitchen sink views with clones 1`] = `
Array [
  Object {
    "name": "someView",
    "rules": Array [
      Object {
        "declarations": Array [
          Object {
            "src": Object {
              "args": Array [
                Object {
                  "$ref": "someValue",
                },
                Object {
                  "$ref": "hello",
                },
              ],
              "op": "bwand",
            },
            "targetType": "content",
            "type": "Bind",
          },
          Object {
            "actions": Array [
              Object {
                "event": "someEvent",
                "payload": null,
                "target": "someController",
                "type": "Dispatch",
              },
              Object {
                "type": "PreventDefault",
              },
              Object {
                "type": "StopPropagation",
              },
            ],
            "eventType": "mousedown",
            "type": "DOMEvent",
          },
          Object {
            "actions": Array [
              Object {
                "event": "hello",
                "payload": null,
                "target": "someOtherController",
                "type": "Dispatch",
              },
            ],
            "eventType": "click",
            "type": "DOMEvent",
          },
        ],
        "selector": "#some-id",
      },
      Object {
        "declarations": Array [
          Object {
            "childRules": Array [
              Object {
                "declarations": Array [
                  Object {
                    "src": Object {
                      "$ref": "value",
                    },
                    "targetType": "content",
                    "type": "Bind",
                  },
                  Object {
                    "src": Object {
                      "args": Array [
                        Object {
                          "args": Array [
                            Object {
                              "$ref": "id",
                            },
                            Object {
                              "$primitive": "-class",
                            },
                          ],
                          "op": "array",
                        },
                        Object {
                          "$primitive": "",
                        },
                      ],
                      "op": "join",
                    },
                    "target": undefined,
                    "targetType": "attribute",
                    "type": "Bind",
                  },
                ],
                "selector": ".something",
              },
            ],
            "iterator": Array [
              "id",
              "value",
            ],
            "type": "Clone",
          },
        ],
        "selector": "#some-clone",
      },
    ],
    "type": "View",
  },
]
`;
